lib/common/pool.h

```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright (c) 2016-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+#ifndef POOL_H
+#define POOL_H
+
+#include <stddef.h>   /* size_t */
+
+typedef struct POOL_ctx_s POOL_ctx;
+
+/*! POOL_create() :
+    Create a thread pool with at most `numThreads` threads.
+    `numThreads` must be at least 1.
+    The maximum number of queued jobs before blocking is `queueSize`.
+    `queueSize` must be at least 1.
+    @return : The POOL_ctx pointer on success else NULL.
+*/
+POOL_ctx *POOL_create(size_t numThreads, size_t queueSize);
+
+/*! POOL_free() :
+    Free a thread pool returned by POOL_create().
+*/
+void POOL_free(POOL_ctx *ctx);
+
+/*! POOL_function :
+    The function type that can be added to a thread pool.
+*/
+typedef void (*POOL_function)(void *);
+/*! POOL_add_function :
+    The function type for a generic thread pool add function.
+*/
+typedef void (*POOL_add_function)(void *, POOL_function, void *);
+
+/*! POOL_add() :
+    Add the job `function(opaque)` to the thread pool.
+    Possibly blocks until there is room in the queue.
+*/
+void POOL_add(void *ctx, POOL_function function, void *opaque);
```

Comment:
A comment : 
when invoking `POOL_add()`, one has to make sure that `opaque` argument will outlive the function call.

It may feel obvious, but it's also easy to make this mistake : 
define `void* opaque` as a kind of `structJob* jobDescription` which is defined on the fly into a `POOL_add()` loop, hence becomes invalid on reaching the end of the loop.

This lifetime issue is an important reason why previous version of zstdmt would use a non-generic job post / consume mechanism : it allowed passing the job description "by value", eliminating lifetime issue. 
Unfortunately, by definition, a generic threadPool cannot do that, since it cannot guess the content of `opaque`.

Ensuring that `void* opaque` outlive its thread requires a more complex strategy which typically collects back `opaque` on job completion.
Nothing earth shattering, but nonetheless something to worry about.
Might be worth a code comment.